tests/instruments/ametek/test_ametek7270.py

```diff
@@ -0,0 +1,526 @@
+import pytest
+
+from pymeasure.test import expected_protocol
+from pymeasure.instruments.ametek import Ametek7270
+
+
+def test_init():
+    with expected_protocol(
+            Ametek7270,
+            [],
+    ):
+        pass  # Verify the expected communication.
+
+
+@pytest.mark.parametrize("comm_pairs, value", (
+    ([(b'SEN', b'27\n')],
+     1.0),
+    ([(b'SEN', b'27\n')],
+     1.0),
+))
+def test_sensitivity_getter(comm_pairs, value):
+    with expected_protocol(
+            Ametek7270,
+            comm_pairs,
+    ) as inst:
+        assert inst.sensitivity == value
+
+
+def test_slope_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'SLOPE', b'1\n')],
+    ) as inst:
+        assert inst.slope == 12
+
+
+def test_time_constant_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'TC', b'30\n')],
+    ) as inst:
+        assert inst.time_constant == 100000.0
+
+
+def test_x_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'X.', b'0.0E+00\n')],
+    ) as inst:
+        assert inst.x == 0.0
+
+
+def test_y_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'Y.', b'0.0E+00\n')],
+    ) as inst:
+        assert inst.y == 0.0
+
+
+def test_x1_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'X1.', b'')],
+    ) as inst:
+        assert inst.x1 == ''
+
+
+def test_y1_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'Y1.', b'')],
+    ) as inst:
+        assert inst.y1 == ''
+
+
+def test_x2_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'X2.', b'')],
+    ) as inst:
+        assert inst.x2 == ''
+
+
+def test_y2_getter():
+    with expected_protocol(
+            Ametek7270,
+            [(b'Y2.', b'')],
+    ) as inst:
+        assert inst.y2 == ''
```

Comment:
well it depends on the lockin mode (single reference or dual harmonic). In single reference mode, only x, y are read as floats all properties indexed with 1 and 2 are empty string and it is the opposite in dual harmonic mode. 

So how do you expect me to write tests in such a situation?

Reply:
With the tests, I realized the type of the return value, so we are talking about the implementation itself.

I would not expect an empty string as a response value, but 0, None, or nan or that an exception is raised.

As it is an invalid query, an exception, None, or nan can be understood. 

Maybe raising ValueError is the best, as you try to read a property, which is invalid to query right now.

Reply:
What would expect, if a property is not defined?
0 could bot be distinguished from a valid response, so 0 is not a good answer

Reply:
The thing is (and I tested it on the instrument right now), no error is generated from the instrument only empty response. I mean the call on the property is not invalid it is just undefined.

 I'm not sure I understand, I mean I'm not the one writing the response in the implementation but the driver (instrument) is. If it replies, in certain circumstances, with an empty string, what should I do?? Should I add a check_get_error thing and depending on the mode raise a ValueError, is that what you mean? But then it is not more coherent with the instrument driver documentation...

Reply:
These drivers, we are writing, hide some details from the user.
Therefore, we introduce changes, we consider appropriate.
Another instrument, for example, returns the temperature times 10. The python code divides by ten and returns an integer, such that the user gets a temperature in degrees.
So there's no problem deviating from the pure instrument communication.

An empty string is not a good response value for an undefined property.
Imagine a user reads that property and tries to calculate something from it.
Then he will get an obscure error, that you can't add Str to int (for example).

If you raise a meaningful exception, the user will notice immediately, what the problem is.

Maybe the best place is the `read` method.

Can you follow me?

Reply:
yeah get it, I'll try to write something!

Reply:
I looked into the read reimplementation, but how to know where this read comes from and that the empty string response is wrong (in some cases it is not, for instance after a property has been set). Then I thought about the casting, because a float on an empty string raise a ValueError but I saw you discard this error in the values method:

```
 for i, result in enumerate(results):
            try:
                if cast == bool:
                    # Need to cast to float first since results are usually
                    # strings and bool of a non-empty string is always True
                    results[i] = bool(float(result))
                else:
                    results[i] = cast(result)
            except Exception:
                pass  # Keep as string
``` 

and the ckeck_get_error=True will just also call the check_errors

So I tried another thing using the get_process method:

```


    y = Instrument.measurement("Y.",
                               """ Reads the Y value in Volts """,
                               get_process=check_read_not_empty,
                               )
    x1 = Instrument.measurement("X1.",
                                """ Reads the first harmonic X value in Volts """,
                                get_process=check_read_not_empty,
                                )
```

with `check_read_not_empty` is:

```
def check_read_not_empty(value):
    """Called by some properties to check if the reply is not an empty string
    that would mean the properties is currently invalid (probably because the reference mode
    is on single or dual)"""
    if value == '':
        raise ValueError('Invalid response from measurement call')
    else:
        return value
```

I tested this on the instrument, changing dynamically the reference mode, and it works well. What do you think?

Reply:
That last code looks good. For reasons like this (or calculating a Value etc.), we have "get_process".

Maybe you could try to make the error message more meaningful, for example hinting at the possibly wrong mode.

It's also good to document in the instrument's docstring this changing between sets of commands.

The idea is, that a user does not have to look at the manual while using your driver.

Reply:
> alright now how should I modify the tests on these properties that are raising an error? The generator is not catching the error. Do you have a layout for this? I know about pytest.raises(ValueError) but how to write it with the expected protocol?

You can checkout the DSP Signal Recovery lockin amplifier tests:
https://github.com/pymeasure/pymeasure/blob/master/tests/instruments/signalrecovery/test_dspbase.py

Or the TDK lambda tests in my PR branch here:
https://github.com/mcdo0486/pymeasure/tree/dev/tdk_lambda_40-38_80-65/tests/instruments/tdk

The TDK does something kind of similar in the sense that "OK" gets returned by instrument so I needed to use `check_set_errors`. You can see the expected protocol examples there. 

Hope that helps.

Reply:
You can put both with statements into each other:
```python
With pytest.raises():
    With expected_protocol() as inst:
        Pass
```

I'd use a parameterized test (getting all these attributes with "geattr(Inst, Name)") for all the raised exceptions, as they are all the same.

You could make these tests work with returning correct values.


Btw. You said, that the instrument returns a newline char at the end of every message. Why don't you include them in the tests? That way you test, whether your code interprets the messages correctly.

Reply:
> alright now how should I modify the tests on these properties that are raising an error? The generator is not catching the error. Do you have a layout for this? I know about pytest.raises(ValueError) but how to write it with the expected protocol?

I tried throwing a PR to your branch, but probably easiest to give the example code here:

```
def test_x1_getter():
    with pytest.raises(ValueError):
        with expected_protocol(
                Ametek7270,
                [(b'X1.', b'')],
        ) as inst:
            assert inst.x1 == ''
```